### PR TITLE
Refactor notifications, activities and reputation

### PR DESCRIPTION
**Pull request by iHiD**
This refactors common code and ideas in notifications, activities and reputation into one unified system. Once finished it will significantly reduce the amount of code in each of the three areas, and bring them in line with each other to make things more consistent and easy to work with.

I am documenting the work in `app/models/concerns/is_paramaterised_sti.rb` if you'd like to read it to understand the methodology.

So far I've roughly brought notifications and activities in line with one another. I now need to do work to add the caching logic into notifications.

The next stage is to turn the commands that @ErikSchierboom built out into STI models inline with this, and update the tests. The fundamental change here is that we encode the data that is in those commands into models instead, where I feel they better below.

This whole thing is the end of a loooong cycle of changes and refactors to try and make all this code fast and easily cacheable as these areas are prone to lots of `n+1`s, and a realm of difficulties in keeping things in sync as the codebase evolves.

---

Note that the first commit here is really just me trying to sanity check whether this would work for reputation. It's better to read just the second one.

### CODE REVIEW

File: app/commands/user/reputation_token/award_for_pull_request.rb [hunk 1]
```diff
@@ -59,11 +73,11 @@ def pr_id
...
 
-      def author_reputation_reason
-        return :'contributed_code/major' if params[:labels].include?('reputation/contributed_code/major')
-        return :'contributed_code/minor' if params[:labels].include?('reputation/contributed_code/minor')
+      def author_reputation_level
+        return :major if params[:labels].include?('reputation/contributed_code/major')
+        return :minor if params[:labels].include?('reputation/contributed_code/minor')
 
-        :'contributed_code/regular'
+        :regular
```
Comment (iHiD) [inline app/commands/user/reputation_token/award_for_pull_request.rb:80]: @ErikSchierboom `reason` and `level` are now two different things. "Reason" is part of the object (I actually think we can probably get rid of it now, but I'm not 100% so I've left it in). "Level"  is passed in where appropriate.

File: app/commands/user/activity/create.rb [hunk 2]
```diff
@@ -15,7 +15,6 @@ def call
 
       klass.create!(
         user: user,
-        track: params.delete(:track),
```
Comment (iHiD) [inline app/commands/user/activity/create.rb:18]: This happens in the Concern now.

**Comment by iHiD** [comment]
@kntsoriano @Erik Schierboom This is pretty much ready to go now.

It is a very meaty set of changes with lots of metaprogramming so probably not reviewable really, but I want you both to understand how it all works.

There is a big block of comments here that explains the key ideas: https://github.com/exercism/website/blob/9f44251306e191b20e709ab9724191b560633986/app/models/concerns/is_paramaterised_sti.rb#L1
But basically, the whole thing defines a DSL for us to use with notifications, activity and reputation. Here is an example of each to look at:
- https://github.com/exercism/website/blob/9f44251306e191b20e709ab9724191b560633986/app/models/user/activities/completed_exercise_activity.rb
- https://github.com/exercism/website/blob/9f44251306e191b20e709ab9724191b560633986/app/models/user/notifications/mentor_started_discussion_notification.rb
- https://github.com/exercism/website/blob/9f44251306e191b20e709ab9724191b560633986/app/models/user/reputation_tokens/code_contribution_token.rb


There are three bits of data that are key to understand:
1. `params`: These are stored in a JSON column in the database. If we use objects, they’re stored using their GlobalID identifiers
2. `cachable_rendering_data` (and its friend `uncachable_rendering_data`): This is where the rendering data is defined. It is combined into `rendering_data`, which is what’s publicly accessed. That data is then what is passed to React. The key thing here is that most of it is able to be cached within the database thus avoiding n+1s at runtime. We don’t cache AR-objects here, only things that can be represented in JSON.
3. `i18n_params`: determines what is passed into the i18n rendering.

Everything else in that PR is basically building the DSL and background functionality that makes all that work, and updating existing code.

Defining new things (notifications, reputation, etc) is now very easy. Basically just copy one of the existing classes, updating whichever params are required for that. And it will all just work (maybe :wink:)

**Review by kntsoriano — APPROVED** [review]

File: app/models/concerns/is_paramaterised_sti.rb [hunk 3]
```diff
@@ -0,0 +1,169 @@
+# ** Introduction  **
+#
+# This class allows for the creation of paramaterised
+# STI classes, with guards for uniquness, i18n versioning
```
Comment (kotp) [inline app/models/concerns/is_paramaterised_sti.rb:4]:
```suggestion
# This class allows for the creation of parameterized
# STI classes, with guards for uniqueness, i18n versioning
```

File: app/commands/user/reputation_token/award_for_pull_request.rb (hunk 1 above)
Comment (ErikSchierboom) [reply app/commands/user/reputation_token/award_for_pull_request.rb:80]: 👍

File: app/models/concerns/is_paramaterised_sti.rb [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+# and cachablity. This powers notifications, user activities
```
Comment (ErikSchierboom) [inline app/models/concerns/is_paramaterised_sti.rb]:
```suggestion
# and cachability. This powers notifications, user activities
```

**Review by ErikSchierboom — APPROVED** [review]

**Pull request merged**